reliability-v2/tasks/Tasks.py

```diff
@@ -0,0 +1,245 @@
+import logging
+import random
+from time import sleep
+import requests
+import sys
+from threading import Lock
+import asyncio
+from tasks.GlobalData import global_data
+from utils.cli import oc,kubectl,shell
+from utils.LoadApp import loadApp
+from integrations.SlackIntegration import slackIntegration
+from concurrent.futures import ThreadPoolExecutor
+
+class Tasks:
+    def __init__(self):
+        self.logger = logging.getLogger('reliability')
+        self.result_lock = Lock()
+        self.results = {}
+
+    def get_results(self):
+        results = ["[Function]: Total / Passed / Failed"]
+        for key,value in self.results.items():
+            passed = value["passed"]
+            failed = value["failed"]
+            total = passed + failed
+            results.append(f"[{key}]: {total} / {passed} / {failed}")
+        results = "\n".join(results)
+        self.logger.info(f"Reliability test results:\n"+ results)
+        return results
+
+    def __log_result(self,rc):
+        # func is the name of the caller function
+        func = sys._getframe(1).f_code.co_name
+        # set default if not already exist
+        self.results.setdefault(func, {"passed":0,"failed":0})
+            
+        with self.result_lock:
+            if rc == 0:
+                self.results[func]["passed"] += 1
+            else:
+                self.results[func]["failed"] += 1
+                    
+    # delete all project for a user
+    def delete_all_projects(self,user):
+        if user == "kubeadmin":
+            return ("admin user not allowed.",1)
+        kubeconfig = global_data.kubeconfigs[user]
+        self.logger.info(f"[Task] User {user}: delete all projects for user {user}")
+        (result,rc) = oc(f"delete project --all",kubeconfig)
+        self.__log_result(rc)
+        return (result,rc)
+
+    # new a project
+    def new_project(self,user,namespace):
+        kubeconfig_admin = global_data.kubeconfigs["kubeadmin"]
+        kubeconfig = global_data.kubeconfigs[user]
+        self.logger.info(f"[Task] User {user}: new project '{namespace}'")
+        retry = 5
+        while retry > 0:
+            result,rc = oc(f"get project {namespace} --no-headers",kubeconfig_admin,ignore_slack=True)
+            if "Terminating" in result:
+                sleep(20)
+                retry -= 1
+            else:
+                # issue during test: in about 1/10 cases, last step gets no project, but
+                # next step shows already exists. Sleep 20s can mitigate but not avoid this issue.
+                sleep(20)
+                break
+        (result,rc) = oc(f"new-project --skip-config-write {namespace}",kubeconfig)
+        if rc == 0:
+            # developer is forbidden to patch resource "namespaces"
+            oc(f"label namespace {namespace} purpose=reliability",kubeconfig_admin)
+        self.__log_result(rc)
+        return (result,rc)
```

Comment:
@mffiedler The error you saw "Error from server (NotFound): namespaces "testuser-4-3" not found" is expected. That is caused by a `oc get project` operation before create a new project. The reason to have this check is as below:
1. After project deleted(previous step), there is a period that the project is in 'Terminating' state. I check and wait until 'Terminating' state is gone.
2. In my test, I found even after 'Terminating' is gone, in about 1/10 case, `oc new-project` still shows project `already exists`, so I always wait 20s after `oc get project`- this can mitigate but not avoid this issue. I met it occasionally in my test.

Reply:
@mffiedler This error is not sent to slack message as a error needs attention, by setting ignore_slack=True. Though harmless you can still see this error in log file. There is a log enhancement in backlog to control what command output to be written to log file with which log level. Then this error will not go to log file.

Reply:
Got it, thanks for the clarification.